src/hotspot/share/cds/archiveBuilder.cpp

```diff
@@ -471,22 +472,27 @@ void ArchiveBuilder::remember_embedded_pointer_in_copied_obj(MetaspaceClosure::R
       // source objects of point_to_it/set_to_null types are not copied
       // so we don't need to remember their pointers.
     } else {
-      if (src_info->read_only()) {
-        _ro_src_objs.remember_embedded_pointer(src_info, ref);
-      } else {
-        _rw_src_objs.remember_embedded_pointer(src_info, ref);
-      }
+      remember_embedded_pointer_in_gathered_obj(src_info, ref->addr());
     }
   }
 }
 
+void ArchiveBuilder::remember_embedded_pointer_in_gathered_obj(SourceObjInfo* src_info, address* field_addr) {
+  if (src_info->read_only()) {
+    _ro_src_objs.remember_embedded_pointer(src_info, field_addr);
+  } else {
+    _rw_src_objs.remember_embedded_pointer(src_info, field_addr);
+  }
+}
+
 void ArchiveBuilder::gather_source_objs() {
   ResourceMark rm;
   log_info(cds)("Gathering all archivable objects ... ");
   gather_klasses_and_symbols();
   GatherSortedSourceObjs doit(this);
-  iterate_sorted_roots(&doit, /*is_relocating_pointers=*/false);
+  iterate_sorted_roots(&doit);
   doit.finish();
+  FileMapInfo::remember_embedded_pointers();
```

Comment:
What is the above call for?
I thought `FileMapInfo::metaspace_pointers_do()` already called from `iterate_roots()` which is called from `iterate_sorted_roots()`.

Reply:
This call was a hack. It was necessary because `SharedClassPathEntry` was not a regular `MetaspaceObj` type. 

I thought rewriting `SharedClassPathEntry` to subclass from `MetaspaceObj` would be too complicated, and wanted to delay that in a future REF. However, it turned out to be fairly straight-forward. I was able to further simplify the code. See commit [46aaec7](https://github.com/openjdk/jdk/pull/13854/commits/46aaec77036982e1454203e8466ccf04d5ff3575)

Now the pointers inside them are remembered during `ArchiveBuilder::gather_source_objs()`, just like other types of objects.